### PR TITLE
Core20

### PR DESCRIPTION
**Pull request by sd-hd**
some changes for core20

### CODE REVIEW

File: snapcraft/internal/meta/application.py [hunk 1]
```diff
@@ -85,7 +85,7 @@ def can_use_wrapper(self, base: Optional[str]) -> bool:
         if self.command_chain:
             return False
 
-        # We only allow wrappers for core and core18.
+        # We only allow wrappers for core, core18 and core20.
```
Comment (sergiusens) [inline snapcraft/internal/meta/application.py]: this is not desired, wrappers are supposed to be gone moving forward.

File: snapcraft/internal/build_providers/_lxd/_images.py [hunk 2]
```diff
@@ -26,11 +26,19 @@ def get_image_source(*, base: str) -> Dict[str, str]:
     :returns: a dictionary with a valid source for PyLXD.
     :raises KeyError: when a non supported base is used.
     """
+
```
Comment (sergiusens) [inline snapcraft/internal/build_providers/_lxd/_images.py]: why not treat the exception in lxd and switch to the devel stream? Makes it future proof. I do not recommend this change though as we are moving to the same images used for launchpad builds moving forward.

File: snapcraft/internal/meta/application.py [hunk 3]
```diff
@@ -27,7 +27,7 @@
 
 
 _COMMAND_ENTRIES = ["command", "stop-command"]
-_MASSAGED_BASES = ["core", "core18"]
+_MASSAGED_BASES = ["core", "core18", "core20"]
```
Comment (sergiusens) [inline snapcraft/internal/meta/application.py]: core20 should no be massaged

File: snapcraft/internal/pluginhandler/_patchelf.py [hunk 4]
```diff
@@ -49,7 +49,7 @@ def __init__(
         :param confinement: the confinement value the snapcraft project is
                             using (i.e.; devmode, strict, classic).
         :param core_base: the base the snap will use during runtime
-                          (e.g.; core, core18).
+                          (e.g.; core, core18, core20).
```
Comment (sergiusens) [inline snapcraft/internal/pluginhandler/_patchelf.py]: this change is not really necessary, it is an e.g. after all.

File: snapcraft/project/_project_options.py [hunk 5]
```diff
@@ -102,14 +102,15 @@
...
+_HOST_CODENAME_FOR_BASE = {"core20": "focal", "core18": "bionic", "core": "xenial"}
 _HOST_COMPATIBILITY = {
     "xenial": ["trusty", "xenial"],
     "bionic": ["trusty", "xenial", "bionic"],
+    "focal" : ["xenial", "bionic"] 
 }
 _STATIC_BASES = ["bare"]
 
-_LINKER_VERSION_FOR_BASE = {"core18": "2.27", "core": "2.23"}
+_LINKER_VERSION_FOR_BASE = {"core20": "2.33", "core18": "2.27", "core": "2.23"}
```
Comment (sergiusens) [inline snapcraft/project/_project_options.py]: this might change, this logic should be revisited for core20. Considering only people using destructive-mode will see it, the entire logic here seems unnecessary moving forward.

File: snapcraft/plugins/meson.py [hunk 6]
```diff
@@ -75,7 +75,7 @@ def __init__(self, name, options, project):
         self.mesonbuilddir = os.path.join(self.builddir, self.snapbuildname)
 
     def _setup_base_tools(self, base):
-        if base in ("core", "core16", "core18"):
+        if base in ("core", "core16", "core18", "core20"):
```
Comment (sergiusens) [inline snapcraft/plugins/meson.py]: can this be in a separate PR?

File: snapcraft/internal/build_providers/_base_provider.py [hunk 7]
```diff
@@ -411,7 +411,7 @@ def _get_home_directory(self) -> pathlib.Path:
 
     def _base_has_changed(self, base: str, provider_base: str) -> bool:
         # Make it backwards compatible with instances without project info
-        if base == "core18" and provider_base is None:
+        if base in ("core18", "core20") and provider_base is None:
```
Comment (sergiusens) [inline snapcraft/internal/build_providers/_base_provider.py]: this should not be needed, this is transitional logic, newer projects should not expect to see this.

**Comment by sd-hd** [comment]
> Hi there, thanks for the effort. I do have a couple of comments though, which is expected as the enablement of core20 in snapcraft is not an easy one! :-)

Hi sergiussens,

thank you :-)

My Problem last year was, that i have no information about the future plans of core20 and snapcraft.

For my development on a snap, was it enough.

I will change your comments soon. ;-)

Thank you

File: snapcraft/plugins/meson.py (hunk 6 above)
Comment (sd-hd) [reply snapcraft/plugins/meson.py]: done

File: snapcraft/internal/build_providers/_lxd/_images.py (hunk 2 above)
Comment (sd-hd) [reply snapcraft/internal/build_providers/_lxd/_images.py]: I accidently removed the history.

The change was:

```
-_BASE_IMAGE = dict(core="16.04", core16="16.04", core18="18.04")                                                                                                                                                                            
+_BASE_IMAGE = dict(core="16.04", core16="16.04", core18="18.04", core20="20.04")
```

File: snapcraft/internal/meta/application.py (hunk 3 above)
Comment (sd-hd) [reply snapcraft/internal/meta/application.py]: reverted

File: snapcraft/internal/build_providers/_base_provider.py (hunk 7 above)
Comment (sd-hd) [reply snapcraft/internal/build_providers/_base_provider.py]: Can't remember excatly where i needed this. 

But i think it was on building the core20. 

There i switched the base to core18 and got an error that the base is None. (or so...)

File: snapcraft/internal/pluginhandler/_patchelf.py (hunk 4 above)
Comment (sd-hd) [reply snapcraft/internal/pluginhandler/_patchelf.py]: reverted

**Comment by sd-hd** [comment]
@sergiusens 
just cleaned up the definitely not needed things.

The still opened conversations. I don't know what todo. ;-)

File: snapcraft/internal/meta/application.py (hunk 1 above)
Comment (sd-hd) [reply snapcraft/internal/meta/application.py]: Are the wrappers removed immediately?
Or is there a backward compatibility?

**Comment by sergiusens** [comment]
Closing this as core20 is taken care of by PluginV2.
Thanks for the contribution.
